### PR TITLE
BUG: DataFrameGroupBy.value_counts() fails if as_index=False and there are duplicate column labels

### LINKED ISSUE #44992: BUG: `DataFrameGroupBy.value_counts()` fails if `as_index=False` and there are duplicate column labels
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
df = pd.DataFrame([[1, 1, 1]], columns=["a", "b", "b"])
df.groupby("a", as_index=False).value_counts()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_16680/431465574.py in <module>
----> 1 df.groupby("a", as_index=False).value_counts()

c:\users\john\onedrive\documents\github\pandas_johnzangwill\pandas\core\groupby\generic.py in value_counts(self, subset, normalize, sort, ascending, dropna)
   1756             if not self.as_index:
   1757                 # Convert to frame
-> 1758                 result = result.reset_index(name="proportion" if normalize else "count")
   1759             return result.__finalize__(self.obj, method="value_counts")
   1760 

c:\users\john\onedrive\documents\github\pandas_johnzangwill\pandas\util\_decorators.py in wrapper(*args, **kwargs)
    309                     stacklevel=stacklevel,
    310                 )
--> 311             return func(*args, **kwargs)
    312 
    313         return wrapper

c:\users\john\onedrive\documents\github\pandas_johnzangwill\pandas\core\series.py in reset_index(self, level, drop, name, inplace)
   1489 
   1490             df = self.to_frame(name)
-> 1491             return df.reset_index(level=level, drop=drop)
   1492 
   1493     # ----------------------------------------------------------------------

c:\users\john\onedrive\documents\github\pandas_johnzangwill\pandas\util\_decorators.py in wrapper(*args, **kwargs)
    309                     stacklevel=stacklevel,
    310                 )
--> 311             return func(*args, **kwargs)
    312 
    313         return wrapper

c:\users\john\onedrive\documents\github\pandas_johnzangwill\pandas\core\frame.py in reset_index(self, level, drop, inplace, col_level, col_fill)
   5832                     )
   5833 
-> 5834                 new_obj.insert(0, name, level_values)
   5835 
   5836         new_obj.index = new_index

c:\users\john\onedrive\documents\github\pandas_johnzangwill\pandas\core\frame.py in insert(self, loc, column, value, allow_duplicates)
   4433         if not allow_duplicates and column in self.columns:
   4434             # Should this be a different kind of error??
-> 4435             raise ValueError(f"cannot insert {column}, already exists")
   4436         if not isinstance(loc, int):
   4437             raise TypeError("loc must be int")

ValueError: cannot insert b, already exists
```

### Issue Description

When a `DataFrame` is grouped with `as_index=False`, then `value_counts()` returns a similar `DataFrame` with some of the same columns and an additional result column. 
This should still work even if the original `DataFrame` has duplicate column labels. 

### Expected Behavior

```
df.groupby("a", as_index=False).value_counts()
   a  b  b  count
0  1  1  1      1
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 2ab1d1f8d8ecc2b62b4dc2d5c7da8f68f750527c
python           : 3.8.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19043
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United Kingdom.1252

pandas           : 1.4.0.dev0+1433.g2ab1d1f8d8
numpy            : 1.21.4
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 59.4.0
Cython           : 0.29.25
pytest           : 6.2.5
hypothesis       : 6.31.4
sphinx           : 4.3.1
blosc            : None
feather          : None
xlsxwriter       : 3.0.2
lxml.etree       : 4.6.4
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 7.30.1
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : 1.3.2
fsspec           : 2021.11.0
fastparquet      : 0.7.2
gcsfs            : 2021.11.0
matplotlib       : 3.5.0
numexpr          : 2.8.0
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 6.0.1
pyxlsb           : None
s3fs             : 2021.11.0
scipy            : 1.7.3
sqlalchemy       : 1.4.28
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by johnzangwill**
- [x] closes #44992
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

DataFrameGroupBy.value_counts() now succedes if there are duplicate column labels. In addition, it fails if `as_index=False` and column labels clash with the results column, either "count" or "proportion".

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1750,13 +1753,32 @@ def value_counts(
...
                 # Convert to frame
-                result = result.reset_index(name="proportion" if normalize else "count")
+                name = "proportion" if normalize else "count"
+                columns = result.index.names
+                if name in columns:
+                    raise ValueError(
+                        f"Column label '{name}' is duplicate of result column"
+                    )
+                columns = com.fill_missing_names(columns)
+                values = result.values
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:1771]: result is a Series at this point?  ._values is generally preferable to .values, as the latter will cast dt64tz to ndarray[object]

Comment (johnzangwill) [reply pandas/core/groupby/generic.py]: Thanks!
I decided to do this differently following a suggestion from a while ago (https://github.com/pandas-dev/pandas/pull/44755#issuecomment-986258740)
I really wanted these methods to be public, but it is not happening...
Anyway, this way is more DRY and I shouldn't really be writing my own version of reset_index().

**Review by jreback — CHANGES_REQUESTED** [review]
pls don't reinvent the wheel

**Comment by johnzangwill** [comment]
> pls don't reinvent the wheel

I don't know what you mean. I haven't reinvented anything. I just added private methods with an additional parameter as was suggested in https://github.com/pandas-dev/pandas/pull/44755#issuecomment-986258740. No new code, no change to public API. 
Otherwise, please tell me how you would suggest that this is done.

File: pandas/core/series.py [hunk 2]
```diff
@@ -1462,6 +1462,19 @@ def reset_index(self, level=None, drop=False, name=lib.no_default, inplace=False
...
+    def _reset_index(
+        self,
+        level: Hashable | Sequence[Hashable] | None = None,
+        drop: bool = False,
+        name: Hashable | lib.NoDefault = lib.no_default,
+        inplace: bool = False,
+        allow_duplicates: bool = False,
+    ) -> Series | DataFrame | None:
+        """
+        Private version of reset_index with additional allow_duplicates parameter
```
Comment (jbrockmendel) [inline pandas/core/series.py:1476]: the extra paramater is ignored for Series right?

Comment (jbrockmendel) [reply pandas/core/series.py:1476]: nope, never mind

File: pandas/tests/groupby/test_frame_value_counts.py [hunk 3]
```diff
@@ -406,33 +406,51 @@ def test_mixed_groupings(normalize, expected_label, expected_values):
...
 @pytest.mark.parametrize("as_index", [False, True])
-def test_column_name_clashes(test, expected_names, as_index):
-    df = DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8], "e": [9, 10]})
-    if test == "repeat":
-        df.columns = list("abbde")
-    else:
-        df.columns = list("abcd") + ["level_1"]
-
+def test_column_label_duplicates(test, columns, expected_names, as_index):
+    # Test for duplicate input column labels and generated duplicate labels
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_frame_value_counts.py:417]: can you add  GH ref

Comment (johnzangwill) [reply pandas/tests/groupby/test_frame_value_counts.py:418]: Done

**Comment by jreback** [comment]
> > pls don't reinvent the wheel
> 
> I don't know what you mean. I haven't reinvented anything. I just added private methods with an additional parameter as was suggested in [#44755 (comment)](https://github.com/pandas-dev/pandas/pull/44755#issuecomment-986258740). No new code at all and no change to the public API. Otherwise, please tell me how you would suggest that this is done.

adding private methods just explodes things. you can simply use concat internally (e.g. concat the index to the frame itself).

**Comment by johnzangwill** [comment]
> > > pls don't reinvent the wheel
> > 
> > 
> > I don't know what you mean. I haven't reinvented anything. I just added private methods with an additional parameter as was suggested in [#44755 (comment)](https://github.com/pandas-dev/pandas/pull/44755#issuecomment-986258740). No new code at all and no change to the public API. Otherwise, please tell me how you would suggest that this is done.
> 
> adding private methods just explodes things. you can simply use concat internally (e.g. concat the index to the frame itself).

I did not want  to use private methods, nor to reinvent the wheel! 
As you know, I wanted to add `allow_duplicates` to `reset_index()`. That would make this a one liner.

So now this is basically a reinvented `reset_index()`.
BTW I could not use `Series.to_frame()` because:
- It throws away duplicate level names
- MultiIndex has thrown away some dtypes, so I need to set each column

I could have used `concat`, but then I would have had to add the name and the index and then remove the index at the end. `insert` was simpler and neater.

Perhaps this case illuminates the discussion regarding column label duplicates. Here, the caller wants fine control over the duplicate creation policy, irrespective of any overall setting in the dataframe flags.

File: pandas/core/groupby/generic.py [hunk 4]
```diff
@@ -1754,10 +1757,27 @@ def value_counts(
...
                 )
 
-            if not self.as_index:
+            if self.as_index:
+                return result.__finalize__(self.obj, method="value_counts")
+            else:
                 # Convert to frame
-                result = result.reset_index(name="proportion" if normalize else "count")
-            return result.__finalize__(self.obj, method="value_counts")
+                name = "proportion" if normalize else "count"
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1764]: i dont' understand why you are reinventing all kinds of code here. can you not just concat?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 4 above)
Comment (johnzangwill) [reply pandas/core/groupby/generic.py:1764]: I tried! AFAIK concat does not do what is required here. I eplained why not in my comments above:

> I could not use `Series.to_frame()` because:
> 
> * It throws away duplicate level names
> * MultiIndex has thrown away some dtypes, so I need to set each column
> 
> I could have used `concat`, but then I would have had to add the name and the index and then remove the index at the end. `insert` was simpler and neater.

**Comment by johnzangwill** [comment]
@jreback 
> concat the index to the frame itself

> can you not just concat?

- concat does not take an Index argument.
- Index.to_frame does not handle repeated  labels.
- MultiIndex does not handle bool dtype.
- ...

So I think that my way is optimal, but I am always open to suggestions. Perhaps fixing Index.to_frame as a separate PR would be an option. But I would still need to go through the columns fixing the dtypes (as reset_index does)

**Comment by jbrockmendel** [comment]
> MultiIndex does not handle bool dtype.

#45061 introduces Index[bool].  Would that get us within sight of a solution?

**Comment by johnzangwill** [comment]
> #45061 introduces Index[bool]. Would that get us within sight of a solution?

That would help by reducing many of the calls to `lib.maybe_convert_objects` scattered around the system. It would remove a few lines from this PR. 

IMO the "correct" solution to this PR is my #44755, whch currently seems irretrievably blocked due to the `allow_duplicates` parameter. Then it would be just one line: `result.reset_index(allow_duplicates=True)`

@jreback demands using concat instead of my code here. But I explained above why that is not working. Fixing MultiIndex.to_frame (#45245) would help, and your #45061 would simplify that slightly.

I am inclined to launch a PR for #45245 but omitting the bool fix, which would go away once #45061 was merged. But I will need to add the dreaded `allow_duplicates` parameter to `MultiIndex.to_frame`...

File: pandas/core/groupby/generic.py [hunk 5]
```diff
@@ -1754,10 +1757,27 @@ def value_counts(
...
+                    )
+                columns = com.fill_missing_names(columns)
+                result_frame = DataFrame()
+                for i, column in enumerate(columns):
+                    level_values = result.index.get_level_values(i)._values
+                    if level_values.dtype == np.object_:
+                        level_values = lib.maybe_convert_objects(
+                            cast(np.ndarray, level_values)
+                        )
+                    result_frame.insert(i, column, level_values, allow_duplicates=True)
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1778]: this almost O(n^2) inefficient.
so what you need to do is
- if we don't have duplicates in the columns names, concat and are done
- if we do, then a) should this really be the answer? why are we not raising?
- if we actually do allow this then you simply change the column names to a range index, concat, then set them again.


pls don't conflate this issue with reset_index or allow_duplicates, they are completely orthogonal.

this is not going to move forward this keeps re-inventing the wheel.

Comment (johnzangwill) [reply pandas/core/groupby/generic.py]: Not O(n^2)! But O(n), and of course I would rather not be itererating through an axis. That is the whole point of Pandas!

concat does not work, as I explained many times. The best I could do was:

```
result_frame = index.set_names(range(len(columns))).to_frame()
result_frame = concat([result_frame, result], axis=1).reset_index(drop=True).set_axis(columns + [name], axis=1)
```

which fails 6 of my tests due to bool/object problems in MultiIndex. These will probably be fixed by #45061, but I see that has been deferred until 1.5.

Meanwhile, I employed your column renaming suggestion. It is now all green (apart from the usual s3 problem)

**Comment by jbrockmendel** [comment]
Looking at this, I don't see any immediately obvious way to make this simpler than what @johnzangwill has done.

IIUC this can be simplified if/when the PRs mentioned [here](https://github.com/pandas-dev/pandas/pull/45160#issuecomment-1007947557) are merged.  If we are optimistic about those going in soon-ish, then this can go on the backburner until then (and still go in for 1.5).  Otherwise, we should merge with this approach and trust @johnzangwill to follow-up with simplifications as they become feasible.

**Comment by johnzangwill** [comment]
> If we are optimistic about those going in soon-ish

@jbrockmendel, I am not optimistic about my #45160 getting merged soon! @jreback strongly disagrees with my approach, which follows suggestions from other reviewers. What I would like would be for the reviewers to sort this out between them!
It could also be the case that #45061 together with my new #45318 would enable this to be completely vectorized using concat. But the correct place would be within reset_index, so it would still need #45160.

File: pandas/core/groupby/generic.py [hunk 6]
```diff
@@ -1754,10 +1754,22 @@ def value_counts(
                     level=index_level, sort_remaining=False
                 )
 
-            if not self.as_index:
+            if self.as_index:
+                return result.__finalize__(self.obj, method="value_counts")
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1758]: do `result_frame = result` and then do the `__finalize__` after the if/then (e.g. share it between these)

File: pandas/core/groupby/generic.py [hunk 7]
```diff
@@ -1754,10 +1754,22 @@ def value_counts(
...
+            else:
                 # Convert to frame
-                result = result.reset_index(name="proportion" if normalize else "count")
-            return result.__finalize__(self.obj, method="value_counts")
+                name = "proportion" if normalize else "count"
+                index = result.index
+                columns = com.fill_missing_names(index.names)
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1763]: why is this needed?

File: pandas/core/groupby/generic.py [hunk 8]
```diff
@@ -1754,10 +1754,22 @@ def value_counts(
...
+                if name in columns:
+                    raise ValueError(
+                        f"Column label '{name}' is duplicate of result column"
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1766]: why does this matter at all, simply set the columns

File: pandas/core/groupby/generic.py [hunk 9]
```diff
@@ -1754,10 +1754,22 @@ def value_counts(
...
+                    )
+                result.name = name
+                result.index = index.set_names(range(len(columns)))
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1769]: again this seems way more complicated here. can you not simply
```result_frame.columns = ....```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 9 above)
Comment (johnzangwill) [reply pandas/core/groupby/generic.py:1769]: Done

File: pandas/core/groupby/generic.py (hunk 7 above)
Comment (johnzangwill) [reply pandas/core/groupby/generic.py:1763]: When the groupby "as" arguments are arrays or functions then the index level names come through as None. These have to converted to legitimate column headings for the frame. This is exactly what reset_index does internally.

File: pandas/core/groupby/generic.py (hunk 8 above)
Comment (johnzangwill) [reply pandas/core/groupby/generic.py:1766]: I would have throught that this would invariably be the result of an error. I want value_counts to survive in the case that there were duplicate headings in the original dataframe. In this case, however, there is a clash between one of the original headings and the new heading ("count" or "proportion") being created by value_counts. Other reviewers want cases such as this to raise, since it is clearly unintentional. I will reluctantly take it out, if you insist. Please confirm.

File: pandas/core/groupby/generic.py (hunk 6 above)
Comment (johnzangwill) [reply pandas/core/groupby/generic.py]: Done

File: pandas/core/groupby/generic.py (hunk 7 above)
Comment (jreback) [reply pandas/core/groupby/generic.py:1766]: ok then pls make a private function out of this (on MI or index if needed) and consolidate both cases to call

File: pandas/core/groupby/generic.py (hunk 8 above)
Comment (jreback) [reply pandas/core/groupby/generic.py:1769]: pls point to the test case for this

File: pandas/core/groupby/generic.py (hunk 7 above)
Comment (johnzangwill) [reply pandas/core/groupby/generic.py:1766]: I did. It is com.fill_missing_names, and I changed reset_index to call it some time ago (My #44878)

File: pandas/core/groupby/generic.py (hunk 8 above)
Comment (johnzangwill) [reply pandas/core/groupby/generic.py:1769]: https://github.com/johnzangwill/pandas/blob/207b55ecca5c5f81b3e0f2b337747f8880b4b45f/pandas/tests/groupby/test_frame_value_counts.py#L447

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @johnzangwill
